### ISSUE TITLE
Reschedule the internal timer instead of checking every 2ms?

### ISSUE THREAD

**Issue by sashee** [simenkid/nodemcu-timer#2]
I noticed that your implementation uses an internal timer that fires every 2ms, even in the case when the next timeout is further away. If I invoke a setTimeout 2 seconds from now, it fires a 1000 times until that.

I've been thinking of an implementation that reschedules the internal timer instead of ticking it every 2ms. The module knows when the next event will need to fire and could set the timer to that moment.

For example, if I set 2 timeout, one to 1 sec, and one to 3 secs, the module could set the timer to 1 second, then after it fired, reschedule to 2 seconds, then delete. In case a new timeout is scheduled, it can always modify the timer to fire at the next event.

I think it would help when the ESP module is running from battery.

I'm curious whether you already tried this approach and found out that it's unfeasible for some reason I'm yet to see.

Anyway, thanks for the excellent library!

**Comment by simenkid** [comment]
Hi sashee, thanks for your feedback. Your idea is cool, and I was not trying the approach you mentioned. 

Here is a problem I'm thinking of. If I set two timeout object, say **tobjA** and **tobjB**. Firstly, I set 3 secs to let **tobjA** start ticking away, and next I set **tobjB** with 1 sec, immediately. I am expecting that **tobjB** will fire before **tobjA**. If your approach can _reschedule_ correctly, that would be great! 

Although it may not be easy, I am very inspired by your idea. Let's try to think if there are possibilities to further reduce the internal busy things, e.g., checking the task queue every 2ms.

Here is another thing that interests me. I wanna know how **nodemcu-timer** impacts the power consumption of ESP8266. Maybe I'll find it out when I am free. 

Since ESP modules consumes current around 50mA(average), it is not easy to keep it longlife when using a battery without a sleeping mechanism. I think the most consumption is coming from WiFi RF transceiver, the consumption from MCU itself may be minor.

**Comment by simenkid** [comment]
I forgot one thing. If a 2-ms tick is too small for your application, you can modify the source

``` lua
local tick, lock = 2, false
```

to 

``` lua
local tick, lock = X, false
```

where X is the interval per tick. Let's say if X = 500 (ms), and `setTimeout(callback, 10)` with 10 ticks will fire in 5 seconds. And **nodemcu-timer** will internally check task queue every 500ms, this will keep it from busy checking.

Here is an example:

``` lua
-- fires after (500ms * 10) = 5 seconds
local tobj = timer.setTimeout(function ()
    print('I am fired')
end, 10)
```

**Comment by sashee** [comment]
Hi,

I've looked into the docs, and it seems like the tmr function are not reliable on keeping the time. The tmr.now() overflows every 35 minutes, and tmr.time() has a precision of 1 second. Refs: http://www.esp8266.com/viewtopic.php?p=4217 and http://www.thalin.se/2015/04/nodemcu-tmrtime-and-tmrnow-bugs.html .

We might be able to circumvent the overflow problem by checking every 35 minutes and keeping track how many overflows are needed to reach the target (for example for a setTimout of 80 minutes it's roughly 2 overflows + 10 minutes).

But the second post mentions that he observed cases where the timer just stopped. His conclusion is that a ticking timer is the most reliable source of time for now.

There is a module called RTC time which should be suitable: https://nodemcu.readthedocs.org/en/dev/en/modules/rtctime/#rtctimeget . It's a custom module and not available in the base, but can be compiled in.

These are what I've found so far, I'll have a go and implement something usable in the following days.
